octodns/provider/base.py

```diff
@@ -28,26 +30,19 @@ def __init__(self, id, apply_disabled=False,
         self.apply_disabled = apply_disabled
         self.update_pcent_threshold = update_pcent_threshold
         self.delete_pcent_threshold = delete_pcent_threshold
-
-    def _include_change(self, change):
-        '''
-        An opportunity for providers to filter out false positives due to
-        peculiarities in their implementation. E.g. minimum TTLs.
-        '''
-        return True
-
-    def _extra_changes(self, existing, desired, changes):
-        '''
-        An opportunity for providers to add extra changes to the plan that are
-        necessary to update ancillary record data or configure the zone. E.g.
-        base NS records.
-        '''
-        return []
+        self.strict_supports = strict_supports
 
     def _process_desired_zone(self, desired):
         '''
-        Providers can use this method to make any custom changes to the
-        desired zone.
+        An opportunity for providers to modify that desired zone records before
+        planning.
+
+        - Must do their work and then call super with the results of that work
+        - Must not modify the `desired` parameter or its records and should
+          make a copy of anything it's modifying
```

Comment:
Should we send a copy of `desired` to begin with?

Reply:
Went back and forth on that. We'd have to copy the zone and all the records to make it safe. Decided that was a bunch of overhead for things that 90-whatever percent of the time won't actually make copies, but I could be convinced.

Reply:
We're slowly increasing programability available to the providers. That also starts increasing the chances of providers doing unwanted things and potentially 🦶 🔫 . I guess it's a paradigm choice to make on our part. safety vs simplicity/performance (I am guessing large zones will have non trivial overhead of copying per provider). I don't have a strong opinion either way.

A good balance could be to create a copy only if a provider re-implements the function. My Python OOP is not good enough to recall if this is possible.

Reply:
I'm okay keeping it as-is too and letting providers re-implement it properly.

Reply:
Been thinking about this a bit more. I think we could potentially come up with a setup that does ~copy-on-write. 

I.e. some sort of `Zone.shallow_copy` that could be passed in at fairly low cost, e.g. the `Zone` object and it's set would be duplicated, but all the `Record` objects would be shared. Could then add `Zone.replace_record` that would take the modified record and replace it in the shallow copied zone. There may be other options for replacing with less work on the part of the provider.

That said it would still have the overhead of duplicating and coping everything into the `Zone` for 99% of cases where nothing was actually going to change... For smaller/common zones it's probably NBD, but I'm fairly sure there are people managing zones with many 1000's of records and the solution doesn't seem in balance with the (lack of) problem so as interesting as that'd be I'd probably still land with things behaving as they currently do. 

It might be worth exploring the copy-on-write support in general for the providers to use when they do modify things. It might make it easier for them to do that work. Will see how that goes as we move more things into this paradigm. Definitely not a change for this PR.